### PR TITLE
Space leaks

### PR DESCRIPTION
**Pull request by lambdageek**

### CODE REVIEW

**Comment by lambdageek** [comment]
@evincarofautumn I want to talk about these changes a little bit...

File: src/Language/C/Data/Ident.hs [hunk 1]
```diff
@@ -102,7 +103,7 @@ bits28 = 2^(28::Int)
 --
 -- * for reasons of simplicity the complete lexeme is hashed.
 mkIdent            :: Position -> String -> Name -> Ident
-mkIdent pos s name  = Ident s (quad s) (mkNodeInfo' pos (pos,length s) name)
+mkIdent pos s name  = s `deepseq` Ident s (quad s) (mkNodeInfo' pos (pos,length s) name)
```
Comment (lambdageek) [inline src/Language/C/Data/Ident.hs:106]: We could probably avoid a deepseq dependency by just handwriting the seq instance for lists.

File: src/Language/C/Parser/Lexer.x [hunk 2]
```diff
@@ -302,88 +302,88 @@ label __label__
...
+idkwtok ('_' : '_' : 't' : 'y' : 'p' : 'e' : 'o' : 'f' : []) pos = tok 8 CTokTypeof pos
+idkwtok ('t' : 'y' : 'p' : 'e' : 'o' : 'f' : []) pos = tok 6 CTokTypeof pos
+idkwtok ('_' : '_' : 't' : 'y' : 'p' : 'e' : 'o' : 'f' : '_' : '_' : []) pos = tok 10 CTokTypeof pos
+idkwtok ('u' : 'n' : 'i' : 'o' : 'n' : []) pos = tok 5 CTokUnion pos
+idkwtok ('u' : 'n' : 's' : 'i' : 'g' : 'n' : 'e' : 'd' : []) pos = tok 8 CTokUnsigned pos
+idkwtok ('v' : 'o' : 'i' : 'd' : []) pos = tok 4 CTokVoid pos
+idkwtok ('_' : '_' : 'v' : 'o' : 'l' : 'a' : 't' : 'i' : 'l' : 'e' : []) pos = tok 10 CTokVolatile pos
+idkwtok ('v' : 'o' : 'l' : 'a' : 't' : 'i' : 'l' : 'e' : []) pos = tok 8 CTokVolatile pos
+idkwtok ('_' : '_' : 'v' : 'o' : 'l' : 'a' : 't' : 'i' : 'l' : 'e' : '_' : '_' : []) pos = tok 12 CTokVolatile pos
+idkwtok ('w' : 'h' : 'i' : 'l' : 'e' : []) pos = tok 5 CTokWhile pos
```
Comment (lambdageek) [inline src/Language/C/Parser/Lexer.x:384]: I'm really surprised you have to resort to this.  Have you experimented with just marking `tok` (and perhaps `idkwtok`) with `{-# INLINE #-}`.  It's all local to the lexer, right?

Comment (lambdageek) [reply src/Language/C/Parser/Lexer.x:384]: Possibly `tok` needs to be rewritten as

```haskell
tok :: Int -> (PosLength -> CToken) -> Position -> P CToken
tok len tc = \pos -> return (tc (pos,len))
{-# INLINE tok #-}
```

Comment (evincarofautumn) [reply src/Language/C/Parser/Lexer.x:384]: I haven’t—can try that though. I know it seems like voodoo, I just noticed that the RTS was allocating a lot of closures here, even though (IIRC) this is usually fully applied.

Comment (lambdageek) [reply src/Language/C/Parser/Lexer.x:384]: Well I was wrong Core looks identical if I try the 2-arg `tok` and gratuitous `{-# INLINE #-}` or if I use the original unaltered version.  AFAICT all the Token constructors end up fully applied no matter what. 

Let me try getting Core for your version.

Comment (lambdageek) [reply src/Language/C/Parser/Lexer.x:384]: Interesting, the Core did change after I eta-expanded `idkwtok` but only by floating some let bindings from the CTokIdent/ CTokTyIdent case, below.   I don't see anything that would suggest fewer allocations.

**Pull request closed without merging**